### PR TITLE
UE 5.4 Updates

### PR DESCRIPTION
**Pull request by josharms-disguise**
Changes made to get the plugin working with the preview version of UE 5.4.

### CODE REVIEW

File: Source/RenderStream/Public/RenderStreamLink.h [hunk 1]
```diff
@@ -28,7 +28,7 @@ typedef uint64_t VkDeviceSize;
 typedef struct VkSemaphore_T* VkSemaphore;
 
 #define RS_PLUGIN_NAME "RenderStream-UE"
-#define RS_PLUGIN_VERSION "RS2.0-UE5.3-v2"
+#define RS_PLUGIN_VERSION "RS2.0-UE5.4-v2"
```
Comment (pkevin-disguise) [inline Source/RenderStream/Public/RenderStreamLink.h]: this should probably be: RS2.0-UE5.4-v1

**Review by pkevin-disguise — APPROVED** [review]
I'm not super familiar with what is required to upgrade a plugin to the next version, but this lgtm - unsure why we needed to remove the CONTEXT() call from the RenderStreamHelper though.

File: Source/RenderStream/Private/RenderStreamViewportClient.cpp [hunk 2]
```diff
@@ -455,11 +683,8 @@ void URenderStreamViewportClient::Draw(FViewport* InViewport, FCanvas* SceneCanv
 
 					View->CameraConstrainedViewRect = View->UnscaledViewRect;
 
-					// Enable per-view virtual shadow map caching
-					View->State->AddVirtualShadowMapCache(MyWorld->Scene);
-
 					// Enable per-view Lumen scene
-					if (DisplayClusterLumenPerView->GetInt())
+					if (GDisplayClusterLumenPerView)
```
Comment (MunWolf) [inline Source/RenderStream/Private/RenderStreamViewportClient.cpp]: This change is bad, you added the globals at the top so now 2 of them will be defined, we should be accessing them through their API as was done before.

File: Source/RenderStream/Private/RenderStreamViewportClient.cpp [hunk 3]
```diff
@@ -653,16 +878,14 @@ void URenderStreamViewportClient::Draw(FViewport* InViewport, FCanvas* SceneCanv
 				ViewFamily.bIsHDR = GetWindow().IsValid() ? GetWindow().Get()->GetIsHDR() : false;
 
 #if WITH_MGPU
-				ViewFamily.bForceCopyCrossGPU = DisplayClusterForceCopyCrossGPU->GetInt() != 0;
+				ViewFamily.bForceCopyCrossGPU = GDisplayClusterForceCopyCrossGPU != 0;
```
Comment (MunWolf) [inline Source/RenderStream/Private/RenderStreamViewportClient.cpp]: Same here

File: Source/RenderStream/Private/RenderStreamViewportClient.cpp [hunk 4]
```diff
@@ -680,13 +903,16 @@ void URenderStreamViewportClient::Draw(FViewport* InViewport, FCanvas* SceneCanv
...
+	IDisplayCluster::Get().GetCallbacks().OnDisplayClusterPreSubmitViewFamilies().Broadcast(ViewFamilies);
+
 	// We gathered all the view families, now render them
 	if (!ViewFamilies.IsEmpty())
 	{
 		if (ViewFamilies.Num() > 1)
 		{
 #if WITH_MGPU
-			if (DisplayClusterSortViews->GetInt())
+			if (GDisplayClusterSortViews)
```
Comment (MunWolf) [inline Source/RenderStream/Private/RenderStreamViewportClient.cpp]: Here

File: Source/RenderStream/Private/RenderStreamViewportClient.cpp [hunk 5]
```diff
@@ -706,7 +932,7 @@ void URenderStreamViewportClient::Draw(FViewport* InViewport, FCanvas* SceneCanv
 			}
 		}
 
-		if (DisplayClusterSingleRender->GetInt())
+		if (GDisplayClusterSingleRender)
```
Comment (MunWolf) [inline Source/RenderStream/Private/RenderStreamViewportClient.cpp]: Here

File: Source/RenderStream/Private/RenderStreamViewportClient.cpp [hunk 6]
```diff
@@ -809,7 +1028,7 @@ void URenderStreamViewportClient::Draw(FViewport* InViewport, FCanvas* SceneCanv
 #endif
 		}
 
-		if (DisplayClusterDebugDraw->GetInt() && !ViewFamilies.IsEmpty())
+		if (GDisplayClusterDebugDraw && !ViewFamilies.IsEmpty())
```
Comment (MunWolf) [inline Source/RenderStream/Private/RenderStreamViewportClient.cpp]: Here

File: Source/RenderStream/Private/SceneSelector_StreamingLevels.cpp [hunk 7]
```diff
@@ -117,7 +121,7 @@ bool SceneSelector_StreamingLevels::ValidateLevel(uint32_t sceneId)
     RenderStreamLink::RemoteParameters& parameters = Schema().scenes.scenes[sceneId];
     const SchemaSpec& spec = m_specs[sceneId];
     UE_LOG(LogRenderStream, Log, TEXT("SceneSelectorStreamingLevels: Validating schema for %s with %d parameters"), UTF8_TO_TCHAR(parameters.name), parameters.nParameters);
-    AActor* levelRoot = spec.streamingLevel ? spec.streamingLevel->GetLevelScriptActor() : nullptr;
+    AActor* levelRoot = spec.streamingLevel ? reinterpret_cast<AActor*>(spec.streamingLevel->GetLevelScriptActor()) : nullptr;
```
Comment (MunWolf) [inline Source/RenderStream/Private/SceneSelector_StreamingLevels.cpp]: Why do we not use a static_cast here?

**Review by MunWolf — CHANGES_REQUESTED** [review]
Outside of the globals and that one static cast it looks good.

File: Source/RenderStream/Private/RenderStreamViewportClient.cpp [hunk 8]
```diff
@@ -171,6 +175,65 @@ ULocalPlayer* URenderStreamViewportClient::SetupInitialLocalPlayer(FString& OutE
...
+);
+
+// Replaces FApp::HasFocus
+bool GDisplayClusterReplaceHasFocusFunction = true;
+static FAutoConsoleVariableRef CVarDisplayClusterReplaceHasFocusFunction(
+	TEXT("DC.ReplaceHasFocusFunction"),
+	GDisplayClusterReplaceHasFocusFunction,
+	TEXT("Replaces the function that FApp::HasFocus() uses, to mitigate OS stalls that happen in some systems."),
+	ECVF_ReadOnly
+);
```
Comment (MunWolf) [inline Source/RenderStream/Private/RenderStreamViewportClient.cpp]: By globals I mean these shouldn't exist in this file

File: Source/RenderStream/Private/SceneSelector_StreamingLevels.cpp (hunk 7 above)
Comment (josharms-disguise) [reply Source/RenderStream/Private/SceneSelector_StreamingLevels.cpp]: Initially it was complaining that it couldn't do a static_cast for those types but after I included the header file for the base class in the file it started working.

**Review by MunWolf — APPROVED** [review]
LGTM

**Comment by MunWolf** [comment]
Closing because we don't actually want this in the UE5.3 branch

**Pull request closed without merging**